src/Adapter/Product/Repository/ProductRepository.php

```diff
@@ -357,4 +405,139 @@ public function bulkDelete(array $productIds): void
+            return [];
+        }
+
+        $shops = [];
+        foreach ($result as $shop) {
+            $shops[] = new ShopId((int) $shop['id_shop']);
+        }
+
+        return $shops;
+    }
+
+    /**
+     * This override was needed because of the extra parameter in product constructor
+     *
+     * {@inheritDoc}
+     */
+    protected function constructObjectModel(int $id, string $objectModelClass, ?int $shopId): ObjectModel
+    {
+        return new Product($id, false, null, $shopId);
+    }
+
+    /**
+     * @param ProductId $productId
+     * @param ShopId $shopId
+     *
+     * @return Product
+     *
+     * @throws CoreException
+     */
+    private function getProductByShopId(ProductId $productId, ShopId $shopId): Product
+    {
+        /** @var Product $product */
+        $product = $this->getObjectModelForShop(
+            $productId->getValue(),
+            Product::class,
+            ProductNotFoundException::class,
+            $shopId
+        );
+
+        return $this->loadProduct($product);
+    }
+
+    /**
+     * @param ProductId $productId
+     *
+     * @return Product
+     *
+     * @throws ProductNotFoundException
+     */
+    private function getProductByDefaultShop(ProductId $productId): Product
+    {
+        $defaultShopId = $this->getProductDefaultShopId($productId);
+
+        return $this->getProductByShopId($productId, $defaultShopId);
+    }
+
+    /**
+     * @param Product $product
+     * @param array $propertiesToUpdate
+     *
+     * @throws CoreException
+     * @throws ProductConstraintException
+     * @throws ProductException
+     * @throws ProductPackConstraintException
+     * @throws ProductStockConstraintException
+     * @throws ManufacturerException
+     * @throws TaxRulesGroupException
+     */
+    private function validateProduct(Product $product, array $propertiesToUpdate): void
+    {
+        $taxRulesGroupIdIsBeingUpdated = in_array('id_tax_rules_group', $propertiesToUpdate, true);
+        $taxRulesGroupId = (int) $product->id_tax_rules_group;
+        $manufacturerIdIsBeingUpdated = in_array('id_manufacturer', $propertiesToUpdate, true);
+        $manufacturerId = (int) $product->id_manufacturer;
+
+        if ($taxRulesGroupIdIsBeingUpdated && $taxRulesGroupId !== ProductTaxRulesGroupSettings::NONE_APPLIED) {
+            $this->taxRulesGroupRepository->assertTaxRulesGroupExists(new TaxRulesGroupId($taxRulesGroupId));
+        }
+        if ($manufacturerIdIsBeingUpdated && $manufacturerId !== NoManufacturerId::NO_MANUFACTURER_ID) {
+            $this->manufacturerRepository->assertManufacturerExists(new ManufacturerId($manufacturerId));
+        }
+
+        $this->productValidator->validate($product);
+    }
+
+    /**
+     * @param Product $product
+     *
+     * @return Product
+     *
+     * @throws CoreException
+     */
+    private function loadProduct(Product $product): Product
+    {
+        try {
+            $product->loadStockData();
```

Comment:
loadProduct = loadStockData? What's going on there?

Maybe would be better to have it as "loadRelatedData" or something like this?

Reply:
Yep, if you prefer I don't mind it's true that loadProduct is a bit confusing

Reply:
It's almost like hydrating an object with additional data. Let's try to find a better naming 💪 

cc @PrestaShop/prestashop-maintainers

Reply:
But no need to overthink this one It is specific to Product, it's in a private method and it just loads the stock data (by default the Product object model doesn't). Maybe `loadProductStock` is enough?